lib/ElementHandle.js

```diff
@@ -137,24 +151,58 @@ class ElementHandle extends JSHandle {
     return {x, y, width, height};
   }
 
+  /**
+   * @return {!Promise<?{x: number, y: number, width: number, height: number}>}
+   */
+  async _assertBoundingBox() {
+    const boundingBox = await this.boundingBox();
+    if (!boundingBox)
+      throw new Error('Node is not visible');
+    return boundingBox;
+  }
+
   /**
    *
    * @param {!Object=} options
    * @returns {!Promise<Object>}
    */
   async screenshot(options = {}) {
-    await this._scrollIntoViewIfNeeded();
+    let needsViewportReset = false;
+
+    let boundingBox = await this._assertBoundingBox();
+
+    const viewport = this._page.viewport();
+
+    if (boundingBox.width > viewport.width || boundingBox.height > viewport.height) {
+      const newViewport = {
+        width: Math.max(viewport.width, Math.ceil(boundingBox.width)),
+        height: Math.max(viewport.height, Math.ceil(boundingBox.height)),
+      };
+      await this._page._emulationManager.emulateViewport(this._client, newViewport);
```

Comment:
We try not to use underscore methods from other files. Can you switch this to page.setViewport?

Reply:
Yeah, I wasn't sure about this one. For reference:

```js
  async setViewport(viewport) {
    const needsReload = await this._emulationManager.emulateViewport(this._client, viewport);
    this._viewport = viewport;
    if (needsReload)
      await this.reload();
  }
```

As I understand it, we won't ever cause a reload here, and the `this._viewport` won't actually change unless another operation is running concurrently with `elHandle.screenshot`.

In theory it should be fine, but I'm getting a consistently failing test after making the change.

```
Failures:
1) Page Request.respond should allow mocking binary responses
  Message:
    Node is not visible @ test.js:1564:5
  Stack:
    Error: Node is not visible
        at ElementHandle._assertBoundingBox (/Users/fb/github/brigand/puppeteer/lib/ElementHandle.js:160:13)
        at <anonymous>
        at process._tickCallback (internal/process/next_tick.js:188:7)
```

Reply:
I'm guessing the `tsc` error in CI is because I'm accessing `this._page._emulationManager`.

```
> puppeteer@1.0.0-post tsc /home/travis/build/GoogleChrome/puppeteer
> tsc -p .
lib/ElementHandle.js(181,13): error TS2554: Expected 1 arguments, but got 2.
lib/ElementHandle.js(203,13): error TS2554: Expected 1 arguments, but got 2.
```

Reply:
The tsc error is because you need to rebase.

You should use setViewport, but the reload brings up a good point. You need to copy the other parameters besides width/height into the new viewport. I recommend Object.assign.